The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/27011111844/job/79714786866) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/79714786866) [(plain)](https://github.com/rust-lang/rust/commit/41ee7f1972ae98a4acfcbc0d6c78def03f3e720e/checks/79714786866/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

---- [codegen] tests/codegen-llvm/store-captures.rs stdout ----
------FileCheck stdout------------------------------

------FileCheck stderr------------------------------
/checkout/tests/codegen-llvm/store-captures.rs:10:18: error: CHECK-LABEL: expected string not found in input
 // CHECK-LABEL: define void @store_ptr
                 ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen-llvm/store-captures/store-captures.ll:1:1: note: scanning from here
; ModuleID = 'store_captures.357af163f2ff2dc1-cgu.0'
^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen-llvm/store-captures/store-captures.ll:7:11: note: possible intended match here
define dso_local void @store_ptr(ptr noundef %x, ptr noalias nofree noundef align 4 dereferenceable(4) %y) unnamed_addr #0 {
          ^

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen-llvm/store-captures/store-captures.ll
Check file: /checkout/tests/codegen-llvm/store-captures.rs

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            1: ; ModuleID = 'store_captures.357af163f2ff2dc1-cgu.0' 
label:10'0     X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
            2: source_filename = "store_captures.357af163f2ff2dc1-cgu.0" 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            3: target datalayout = "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-i128:128-n32:64-S128-ni:1:10:20" 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            4: target triple = "wasm32-unknown-wasip1" 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            5:  
label:10'0     ~
            6: ; Function Attrs: nounwind 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
            7: define dso_local void @store_ptr(ptr noundef %x, ptr noalias nofree noundef align 4 dereferenceable(4) %y) unnamed_addr #0 { 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
label:10'1               ?                                                                                                                   possible intended match
            8: start: 
label:10'0     ~~~~~~~
            9:  store ptr %x, ptr %y, align 4 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           10:  ret void 
label:10'0     ~~~~~~~~~~
           11: } 
label:10'0     ~~
           12:  
label:10'0     ~
           13: ; Function Attrs: nounwind 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           14: define dso_local void @store_ref_freeze(ptr noalias nofree noundef readonly align 4 captures(address, read_provenance) dereferenceable(4) %x, ptr noalias nofree noundef align 4 dereferenceable(4) %y) unnamed_addr #0 { 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           15: start: 
label:10'0     ~~~~~~~
           16:  store ptr %x, ptr %y, align 4, !captures !1 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           17:  ret void 
label:10'0     ~~~~~~~~~~
           18: } 
label:10'0     ~~
           19:  
label:10'0     ~
           20: ; Function Attrs: nounwind 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           21: define dso_local void @store_ref_not_freeze(ptr noundef nonnull align 4 %x, ptr noalias nofree noundef align 4 dereferenceable(4) %y) unnamed_addr #0 { 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           22: start: 
label:10'0     ~~~~~~~
           23:  store ptr %x, ptr %y, align 4 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           24:  ret void 
label:10'0     ~~~~~~~~~~
           25: } 
label:10'0     ~~
           26:  
label:10'0     ~
           27: ; Function Attrs: nounwind 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           28: define dso_local void @store_mut_ref(ptr noalias nofree noundef align 4 dereferenceable(4) %x, ptr noalias nofree noundef align 4 dereferenceable(4) %y) unnamed_addr #0 { 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           29: start: 
label:10'0     ~~~~~~~
           30:  store ptr %x, ptr %y, align 4 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           31:  ret void 
label:10'0     ~~~~~~~~~~
           32: } 
label:10'0     ~~
           33:  
label:10'0     ~
           34: ; Function Attrs: nounwind 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           35: define dso_local void @store_mut_ref_as_shared_ref(ptr noalias nofree noundef align 4 dereferenceable(4) %x, ptr noalias nofree noundef align 4 dereferenceable(4) %y) unnamed_addr #0 { 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           36: start: 
label:10'0     ~~~~~~~
           37:  store ptr %x, ptr %y, align 4, !captures !1 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           38:  ret void 
label:10'0     ~~~~~~~~~~
           39: } 
label:10'0     ~~
           40:  
label:10'0     ~
           41: ; Function Attrs: nounwind 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           42: define dso_local void @store_mut_ref_as_ptr(ptr noalias nofree noundef align 4 dereferenceable(4) %x, ptr noalias nofree noundef align 4 dereferenceable(4) %y) unnamed_addr #0 { 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           43: start: 
label:10'0     ~~~~~~~
           44:  store ptr %x, ptr %y, align 4 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           45:  ret void 
label:10'0     ~~~~~~~~~~
           46: } 
label:10'0     ~~
           47:  
label:10'0     ~
           48: ; Function Attrs: nounwind 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           49: define dso_local void @store_slice(ptr noalias nofree noundef nonnull readonly align 4 captures(address, read_provenance) %x.0, i32 noundef range(i32 0, 536870912) %x.1, ptr noalias nofree noundef align 4 dereferenceable(8) %y) unnamed_addr #0 { 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           50: start: 
label:10'0     ~~~~~~~
           51:  store ptr %x.0, ptr %y, align 4, !captures !1 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           52:  %0 = getelementptr inbounds i8, ptr %y, i32 4 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           53:  store i32 %x.1, ptr %0, align 4 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           54:  ret void 
label:10'0     ~~~~~~~~~~
           55: } 
label:10'0     ~~
           56:  
label:10'0     ~
           57: ; Function Attrs: nounwind 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           58: define dso_local void @store_dyn(ptr noundef nonnull %x.0, ptr noalias nofree noundef readonly align 4 captures(address, read_provenance) dereferenceable(20) %x.1, ptr noalias nofree noundef align 4 dereferenceable(8) %y) unnamed_addr #0 { 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           59: start: 
label:10'0     ~~~~~~~
           60:  store ptr %x.0, ptr %y, align 4 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           61:  %0 = getelementptr inbounds i8, ptr %y, i32 4 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           62:  store ptr %x.1, ptr %0, align 4 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           63:  ret void 
label:10'0     ~~~~~~~~~~
           64: } 
label:10'0     ~~
           65:  
label:10'0     ~
           66: attributes #0 = { nounwind "target-cpu"="generic" } 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           67:  
label:10'0     ~
           68: !llvm.ident = !{!0} 
label:10'0     ~~~~~~~~~~~~~~~~~~~~
           69:  
label:10'0     ~
           70: !0 = !{!"rustc version 1.98.0-nightly (db4435142 2026-06-05)"} 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           71: !1 = !{!"address", !"read_provenance"} 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>>>>>>

------------------------------------------

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen-llvm/store-captures/store-captures.ll" "/checkout/tests/codegen-llvm/store-captures.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
stdout: none
--- stderr -------------------------------
/checkout/tests/codegen-llvm/store-captures.rs:10:18: error: CHECK-LABEL: expected string not found in input
 // CHECK-LABEL: define void @store_ptr
                 ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen-llvm/store-captures/store-captures.ll:1:1: note: scanning from here
; ModuleID = 'store_captures.357af163f2ff2dc1-cgu.0'
^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen-llvm/store-captures/store-captures.ll:7:11: note: possible intended match here
define dso_local void @store_ptr(ptr noundef %x, ptr noalias nofree noundef align 4 dereferenceable(4) %y) unnamed_addr #0 {
          ^

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen-llvm/store-captures/store-captures.ll
Check file: /checkout/tests/codegen-llvm/store-captures.rs

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            1: ; ModuleID = 'store_captures.357af163f2ff2dc1-cgu.0' 
label:10'0     X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
            2: source_filename = "store_captures.357af163f2ff2dc1-cgu.0" 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            3: target datalayout = "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-i128:128-n32:64-S128-ni:1:10:20" 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            4: target triple = "wasm32-unknown-wasip1" 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            5:  
label:10'0     ~
            6: ; Function Attrs: nounwind 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
            7: define dso_local void @store_ptr(ptr noundef %x, ptr noalias nofree noundef align 4 dereferenceable(4) %y) unnamed_addr #0 { 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
label:10'1               ?                                                                                                                   possible intended match
            8: start: 
label:10'0     ~~~~~~~
            9:  store ptr %x, ptr %y, align 4 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           10:  ret void 
label:10'0     ~~~~~~~~~~
           11: } 
label:10'0     ~~
           12:  
label:10'0     ~
           13: ; Function Attrs: nounwind 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           14: define dso_local void @store_ref_freeze(ptr noalias nofree noundef readonly align 4 captures(address, read_provenance) dereferenceable(4) %x, ptr noalias nofree noundef align 4 dereferenceable(4) %y) unnamed_addr #0 { 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           15: start: 
label:10'0     ~~~~~~~
           16:  store ptr %x, ptr %y, align 4, !captures !1 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           17:  ret void 
label:10'0     ~~~~~~~~~~
           18: } 
label:10'0     ~~
           19:  
label:10'0     ~
           20: ; Function Attrs: nounwind 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           21: define dso_local void @store_ref_not_freeze(ptr noundef nonnull align 4 %x, ptr noalias nofree noundef align 4 dereferenceable(4) %y) unnamed_addr #0 { 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           22: start: 
label:10'0     ~~~~~~~
           23:  store ptr %x, ptr %y, align 4 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           24:  ret void 
label:10'0     ~~~~~~~~~~
           25: } 
label:10'0     ~~
           26:  
label:10'0     ~
           27: ; Function Attrs: nounwind 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           28: define dso_local void @store_mut_ref(ptr noalias nofree noundef align 4 dereferenceable(4) %x, ptr noalias nofree noundef align 4 dereferenceable(4) %y) unnamed_addr #0 { 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           29: start: 
label:10'0     ~~~~~~~
           30:  store ptr %x, ptr %y, align 4 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           31:  ret void 
label:10'0     ~~~~~~~~~~
           32: } 
label:10'0     ~~
           33:  
label:10'0     ~
           34: ; Function Attrs: nounwind 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           35: define dso_local void @store_mut_ref_as_shared_ref(ptr noalias nofree noundef align 4 dereferenceable(4) %x, ptr noalias nofree noundef align 4 dereferenceable(4) %y) unnamed_addr #0 { 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           36: start: 
label:10'0     ~~~~~~~
           37:  store ptr %x, ptr %y, align 4, !captures !1 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           38:  ret void 
label:10'0     ~~~~~~~~~~
           39: } 
label:10'0     ~~
           40:  
label:10'0     ~
           41: ; Function Attrs: nounwind 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           42: define dso_local void @store_mut_ref_as_ptr(ptr noalias nofree noundef align 4 dereferenceable(4) %x, ptr noalias nofree noundef align 4 dereferenceable(4) %y) unnamed_addr #0 { 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           43: start: 
label:10'0     ~~~~~~~
           44:  store ptr %x, ptr %y, align 4 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           45:  ret void 
label:10'0     ~~~~~~~~~~
           46: } 
label:10'0     ~~
           47:  
label:10'0     ~
           48: ; Function Attrs: nounwind 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           49: define dso_local void @store_slice(ptr noalias nofree noundef nonnull readonly align 4 captures(address, read_provenance) %x.0, i32 noundef range(i32 0, 536870912) %x.1, ptr noalias nofree noundef align 4 dereferenceable(8) %y) unnamed_addr #0 { 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           50: start: 
label:10'0     ~~~~~~~
           51:  store ptr %x.0, ptr %y, align 4, !captures !1 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           52:  %0 = getelementptr inbounds i8, ptr %y, i32 4 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           53:  store i32 %x.1, ptr %0, align 4 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           54:  ret void 
label:10'0     ~~~~~~~~~~
           55: } 
label:10'0     ~~
           56:  
label:10'0     ~
           57: ; Function Attrs: nounwind 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           58: define dso_local void @store_dyn(ptr noundef nonnull %x.0, ptr noalias nofree noundef readonly align 4 captures(address, read_provenance) dereferenceable(20) %x.1, ptr noalias nofree noundef align 4 dereferenceable(8) %y) unnamed_addr #0 { 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           59: start: 
label:10'0     ~~~~~~~
           60:  store ptr %x.0, ptr %y, align 4 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           61:  %0 = getelementptr inbounds i8, ptr %y, i32 4 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           62:  store ptr %x.1, ptr %0, align 4 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           63:  ret void 
label:10'0     ~~~~~~~~~~
           64: } 
label:10'0     ~~
           65:  
label:10'0     ~
           66: attributes #0 = { nounwind "target-cpu"="generic" } 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           67:  
label:10'0     ~
           68: !llvm.ident = !{!0} 
label:10'0     ~~~~~~~~~~~~~~~~~~~~
           69:  
label:10'0     ~
           70: !0 = !{!"rustc version 1.98.0-nightly (db4435142 2026-06-05)"} 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           71: !1 = !{!"address", !"read_provenance"} 
label:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>>>>>>
------------------------------------------

---- [codegen] tests/codegen-llvm/store-captures.rs stdout end ----

```

</details>

--- a/compiler/rustc_codegen_llvm/src/builder.rs
+++ b/compiler/rustc_codegen_llvm/src/builder.rs
@@ -854,6 +854,22 @@ impl<'a, 'll, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
                     self.set_metadata_node(store, llvm::MD_nontemporal, &[one]);
                 }
             }
+            if flags.contains(MemFlags::CAPTURES_READ_ONLY)
+                && crate::llvm_util::get_version() >= (22, 0, 0)
+            {
+                assert!(
+                    self.type_kind(self.val_ty(val)) == TypeKind::Pointer,
+                    "CAPTURED_READ_ONLY is only supported on pointer stores"
+                );
+                let args = [
+                    self.cx.create_metadata(b"address"),
+                    self.cx.create_metadata(b"read_provenance"),
+                ];
+                // FIXME: Switch this to use MD_captures once LLVM 22 is the minimum.
+                let id = self.get_md_kind_id("captures");
+                let md = llvm::LLVMMDNodeInContext2(self.cx.llcx, args.as_ptr(), args.len());
+                self.set_metadata(store, id, md);
+            }
             store
         }
     }

--- a/compiler/rustc_codegen_ssa/src/lib.rs
+++ b/compiler/rustc_codegen_ssa/src/lib.rs
@@ -167,6 +167,10 @@ bitflags::bitflags! {
         const VOLATILE = 1 << 0;
         const NONTEMPORAL = 1 << 1;
         const UNALIGNED = 1 << 2;
+        /// Indicates that writing through the stored pointer is undefined behavior.
+        /// Only valid on stores of pointers, or pairs where the first element is a pointer.
+        /// In the latter case, the flag only applies to the first element of the pair.
+        const CAPTURES_READ_ONLY = 1 << 3;
     }
 }
 

--- a/compiler/rustc_codegen_ssa/src/mir/operand.rs
+++ b/compiler/rustc_codegen_ssa/src/mir/operand.rs
@@ -297,10 +297,20 @@ impl<'a, 'tcx, V: CodegenObject> OperandRef<'tcx, V> {
         bx: &mut Bx,
         dest: PlaceRef<'tcx, V>,
     ) {
+        self.store_with_annotation_and_flags(bx, dest, MemFlags::empty())
+    }
+
+    /// Same as store_with_annotation(), but also specify flags for the store.
+    pub fn store_with_annotation_and_flags<Bx: BuilderMethods<'a, 'tcx, Value = V>>(
+        self,
+        bx: &mut Bx,
+        dest: PlaceRef<'tcx, V>,
+        flags: MemFlags,
+    ) {
         if let Some(instance) = self.move_annotation {
-            bx.with_move_annotation(instance, |bx| self.val.store(bx, dest))
+            bx.with_move_annotation(instance, |bx| self.val.store_with_flags(bx, dest, flags))
         } else {
-            self.val.store(bx, dest)
+            self.val.store_with_flags(bx, dest, flags)
         }
     }
 
@@ -961,7 +971,8 @@ impl<'a, 'tcx, V: CodegenObject> OperandValue<V> {
                 let llptr = bx.inbounds_ptradd(dest.val.llval, bx.const_usize(b_offset.bytes()));
                 let val = bx.from_immediate(b);
                 let align = dest.val.align.restrict_for_offset(b_offset);
-                bx.store_with_flags(val, llptr, align, flags);
+                // The CAPTURES_READ_ONLY flag only applies to the first element.
+                bx.store_with_flags(val, llptr, align, flags & !MemFlags::CAPTURES_READ_ONLY);
             }
         }
     }

--- a/compiler/rustc_codegen_ssa/src/mir/rvalue.rs
+++ b/compiler/rustc_codegen_ssa/src/mir/rvalue.rs
@@ -2,7 +2,7 @@ use itertools::Itertools as _;
 use rustc_abi::{self as abi, BackendRepr, FIRST_VARIANT};
 use rustc_middle::ty::adjustment::PointerCoercion;
 use rustc_middle::ty::layout::{HasTyCtxt, HasTypingEnv, LayoutOf, TyAndLayout};
-use rustc_middle::ty::{self, Instance, Ty, TyCtxt};
+use rustc_middle::ty::{self, Instance, Mutability, Ty, TyCtxt};
 use rustc_middle::{bug, mir, span_bug};
 use rustc_session::config::OptLevel;
 use tracing::{debug, instrument};
@@ -23,7 +23,7 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         rvalue: &mir::Rvalue<'tcx>,
     ) {
         match *rvalue {
-            mir::Rvalue::Use(ref operand, _) => {
+            mir::Rvalue::Use(ref operand, with_retag) => {
                 if let mir::Operand::Constant(const_op) = operand {
                     let val = self.eval_mir_constant(&const_op);
                     if val.all_bytes_uninit(self.cx.tcx()) {
@@ -40,9 +40,20 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                 ) {
                     debug_assert!(!matches!(cg_operand.val, OperandValue::Ref(..)));
                 }
+                // If this is storing a &Freeze reference with a retag, record that it's not
+                // possible to perform writes through the stored pointer.
+                let flags = if let ty::Ref(_, pointee_ty, Mutability::Not) =
+                    operand.ty(self.mir, self.cx.tcx()).kind()
+                    && with_retag.yes()
+                    && pointee_ty.is_freeze(self.cx.tcx(), self.cx.typing_env())
+                {
+                    MemFlags::CAPTURES_READ_ONLY
+                } else {
+                    MemFlags::empty()
+                };
                 // FIXME: consider not copying constants through stack. (Fixable by codegen'ing
                 // constants into `OperandValue::Ref`; why don’t we do that yet if we don’t?)
-                cg_operand.store_with_annotation(bx, dest);
+                cg_operand.store_with_annotation_and_flags(bx, dest, flags);
             }
 
             mir::Rvalue::Cast(

--- a/tests/codegen-llvm/store-captures.rs
+++ b/tests/codegen-llvm/store-captures.rs
@@ -1,0 +1,68 @@
+//@ min-llvm-version: 22
+//@ compile-flags: -O -C no-prepopulate-passes
+
+#![crate_type = "lib"]
+
+use std::cell::UnsafeCell;
+
+#[unsafe(no_mangle)]
+pub fn store_ptr<'a>(x: *const i32, y: &mut *const i32) {
+    // CHECK-LABEL: define {{.*}} @store_ptr
+    // CHECK-NOT: !captures
+    *y = x;
+}
+
+#[unsafe(no_mangle)]
+pub fn store_ref_freeze<'a>(x: &'a i32, y: &mut &'a i32) {
+    // CHECK-LABEL: define {{.*}} @store_ref_freeze
+    // CHECK: store ptr %x, ptr %y, {{.*}}, !captures ![[CAPTURES:[0-9]+]]
+    *y = x;
+}
+
+#[unsafe(no_mangle)]
+pub fn store_ref_not_freeze<'a>(x: &'a UnsafeCell<i32>, y: &mut &'a UnsafeCell<i32>) {
+    // CHECK-LABEL: define {{.*}} @store_ref_not_freeze
+    // CHECK-NOT: !captures
+    *y = x;
+}
+
+#[unsafe(no_mangle)]
+pub fn store_mut_ref<'a>(x: &'a mut i32, y: &mut &'a mut i32) {
+    // CHECK-LABEL: define {{.*}} @store_mut_ref
+    // CHECK-NOT: !captures
+    *y = x;
+}
+
+#[unsafe(no_mangle)]
+pub fn store_mut_ref_as_shared_ref<'a>(x: &'a mut i32, y: &mut &'a i32) {
+    // CHECK-LABEL: define {{.*}} @store_mut_ref_as_shared_ref
+    // CHECK: store ptr %x, ptr %y, {{.*}}, !captures ![[CAPTURES:[0-9]+]]
+    *y = x;
+}
+
+#[unsafe(no_mangle)]
+pub fn store_mut_ref_as_ptr(x: &mut i32, y: &mut *const i32) {
+    // CHECK-LABEL: define {{.*}} @store_mut_ref_as_ptr
+    // CHECK-NOT: !captures
+    *y = x;
+}
+
+#[unsafe(no_mangle)]
+pub fn store_slice<'a>(x: &'a [i32], y: &mut &'a [i32]) {
+    // CHECK-LABEL: define {{.*}} @store_slice
+    // CHECK: store ptr %x.0, ptr %y, {{.*}}, !captures ![[CAPTURES:[0-9]+]]
+    // Second store is slice size, can't have !captures
+    // CHECK-NOT: !captures
+    *y = x;
+}
+
+#[unsafe(no_mangle)]
+pub fn store_dyn<'a>(x: &'a dyn Drop, y: &mut &'a dyn Drop) {
+    // dyn trait is not known Freeze. The vtable could use !captures,
+    // but that's probably not particularly useful.
+    // CHECK-LABEL: define {{.*}} @store_dyn
+    // CHECK-NOT: !captures
+    *y = x;
+}
+
+// CHECK: ![[CAPTURES]] = !{!"address", !"read_provenance"}

